### DISCUSSION TITLE
Run Python from the toolbar icon

### DISCUSSION THREAD

**Discussion by Erimus-Koo** [doublecmd/doublecmd#1715] [General]
I've tried many times on my Mac, and I found that the external command can only set a single executable file: no command line or others.
And even if I set my `.py` as executable and fill it in, it still does not work, maybe because Dbcmd does not know that use which program to run the `.py` file.

Finally, I created a .sh script for execution and invoked Python within this script, which worked.
<img width="741" alt="image" src="https://github.com/doublecmd/doublecmd/assets/17139670/96fbd30d-c8ef-4f59-91ac-7f7cc6a788d9">

I share my `script_dispatcher.sh` below:

```
#!/bin/bash
# 主要用于 Double Commander 的按钮
# 它只能调用一个外部可执行文件，不能直接调用 Python 或者其他程序，所以用一个代理来实现

# 获取所有传递的参数
params="$@"

# 设置常用变量
PYTHON_EXEC="/Users/erimus/.pyenv/shims/python"
DIR_SHARE="PATH TO YOUR PYTHON FILES"

# 使用 osascript 弹出对话框显示参数
show_dialog() {
  local message="$@"
  osascript <<EOF
tell application "System Events"
    activate
    display dialog "$message"
end tell
EOF
}
# show_dialog "Received arguments:" $params

case "$1" in
  archive_file)
    $PYTHON_EXEC "$DIR_SHARE/archive_files.py" "${@:2}"
    ;;
  another_command)
    $PYTHON_EXEC "$DIR_SHARE/another_command.py" "${@:2}"
    ;;
  *)
    show_dialog "Unknown command: $1"
    ;;
esac

```

Perhaps others might find this code snippet useful, so I hope it can be helpful. 
Additionally, I use AppleScript to call my Python script for reusability, and of course, I also work with Windows.

If there's a better way to achieve this, tell me please.

**Comment by Skif-off** [comment]
What about
```
command: /path-to/script.py
params: %t1 archive_file %p
```
?  (Script must be executable) Although this method may not work for the script that creates GUI. Maybe without `.py `.
Or
```
command: /Users/erimus/.pyenv/shims/python
params: %t1 /path-to/script.py %p
```
Or maybe without `%t1`.

P.S. Your method has a plus: one button, but allows to execute different scripts. But drag & drop can be inconvenient.

**Reply by Erimus-Koo** [reply]
I've tried the 2nd method like below, and it worked:
<img width="652" alt="image" src="https://github.com/doublecmd/doublecmd/assets/17139670/e5c59750-09d3-4595-808c-25d0bd2eee4b">
Thanks a lot, it looks better than mine.

The first method won't work because dbcmd doesn't know how to run the Python format, and even if it did, I would need to chmod +x my.py file, which is one extra step.

I still have two questions:

What is the parameter %t1?
Can dbcmd use my .zshrc or other environment settings, and run without the pyenv path /Users/erimus/.pyenv/shims/python?

**Comment by JonnyTech** [comment]
> What is the parameter %t1?

https://doublecmd.github.io/doc/en/variables.html#executeterm

> %t0	Run in terminal and request to close it at the end
> %t1	Run in terminal and stay open at the end

Also it mentions environment variables in that page
